### ISSUE TITLE
Sik-Ho Tang | Review -- Weight Normalization: A Simple Reparameterization to Accelerate Training of Deep Neural Networks.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#109]
Sik-Ho Tang. [Review — Weight Normalization: A Simple Reparameterization to Accelerate Training of Deep Neural Networks](https://sh-tsang.medium.com/review-weight-normalization-a-simple-reparameterization-to-accelerate-training-of-deep-neural-c88fb0918c9b).

**Comment by NorbertZheng** [comment]
## Overview
Weight Normalization: A Simple Reparameterization to Accelerate Training of Deep Neural Networks, Weight Norm, by OpenAI, 2016 NIPS, Over 1400 Citations. Image Classification, CNN, Normalization

 - Weight normalization, a reparameterization of the weight vectors in a neural network that **decouples the length of those weight vectors from their direction**.
 - It speeds up convergence while it does not introduce any dependencies between the examples in a minibatch. This means that weight norm can also be applied successfully
    - to recurrent models such as LSTMs,
    - and to **noise-sensitive applications** (e.g. still noisy even batch is large) such as deep reinforcement learning.

**Comment by NorbertZheng** [comment]
## Weight Normalization
In a standard artificial neural network, the computation of each neuron consists in taking a weighted sum of input features $x$, followed by an elementwise nonlinearity $\Phi$:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/ac280d96-2503-4261-8a25-97f01e0a542f)

Weight normalization re-parameterizes each weight vector $w$ in terms of a parameter vector $v$ and a scalar parameter $g$ and to perform stochastic gradient descent with respect to those parameters instead.

The weight vectors are expressed in terms of the new parameters using:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/90ca5e72-f8b5-4619-a6ab-15d0ec1f4e4d)

where $v$ is a $k$-dimensional vector, $g$ is scalar, and $||v||$ denotes the Euclidean norm of $v$.

And we got $||w||=g$, which is independent of the parameters $v$.

 - **By decoupling the norm of the weight vector $g$ from the direction of the weight vector $\frac{v}{||v||}$, the convergence of stochastic gradient descent optimization is speed up.**

By differentiating the above equation, the gradients of a loss function $L$ with respect to the new parameters $v$, $g$ is obtained:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/0cba4b52-0931-4aa4-90bb-4112c9594bf2)

Unlike with [batch normalization](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651), the expressions above are independent of the minibatch size and thus cause only minimal computational overhead.

**Comment by NorbertZheng** [comment]
## Combining Weight Normalization With Mean-Only [Batch Normalization](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651)
With this normalization method, the minibatch means are subtracted out like with full [batch normalization](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651), but it does not divide by the minibatch standard deviations. That is, neuron activations are computed using:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/9553aa6d-b5d8-4539-9fe6-7f8b4132f120)

where $w$ is the weight vector, parameterized using weight normalization, and $\mu[t]$ is the minibatch mean of the pre-activation $t$.

During training, a running average of the minibatch mean is kept which will be substituted in for $\mu[t]$ at test time.

**Comment by NorbertZheng** [comment]
## Experimental Results
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/3a5b820a-0d83-4975-9de2-b83e0e304ced)
[All-CNN](https://sh-tsang.medium.com/review-striving-for-simplicity-the-all-convolutional-net-2b2f762c1542)-C is used.

Both weight normalization and [batch normalization](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651) provide a significant speed-up over the standard parameterization. [Batch normalization](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651) makes slightly more progress per epoch.

 - However, training with [batch normalization](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651) was about 16% slower compared to the standard parameterization. In contrast, **weight normalization was not noticeably slower**.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/39b25ec2-310d-4f50-9142-c467f51ea073)
Classification results on CIFAR-10 without data augmentation.

Weight normalization, the normal parameterization, and mean-only [batch normalization](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651) have similar test accuracy (~8.5% error). [Batch normalization](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651) does significantly better at 8.05% error.

Mean-only [batch normalization](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651) combined with weight normalization has the best performance at 7.31% test error. It is hypothesized that the noise added by [batch normalization](https://sh-tsang.medium.com/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651) can be useful for regularizing the network.

**Comment by NorbertZheng** [comment]
(There are also experiments on VAE, and deep reinforcement learning in the paper. Please feel free to read if interested.)

**Comment by NorbertZheng** [comment]
## References
 - [2016 NIPS] [Weight Norm] [Weight Normalization: A Simple Reparameterization to Accelerate Training of Deep Neural Networks](https://arxiv.org/abs/1602.07868).
